src/java.base/share/classes/sun/security/ssl/SSLContextImpl.java

```diff
@@ -71,9 +72,42 @@ public abstract class SSLContextImpl extends SSLContextSpi {
     private volatile StatusResponseManager statusResponseManager;
 
     private final ReentrantLock contextLock = new ReentrantLock();
-    final HashMap<Integer,
-            SessionTicketExtension.StatelessKey> keyHashMap = new HashMap<>();
+    private int currentKeyID = new SecureRandom().nextInt();
+    private final Map<Integer,
+            SessionTicketExtension.StatelessKey> keyHashMap = new ConcurrentHashMap<>();
+
+    private void cleanupSessionKeys() {
+        for (Map.Entry<Integer, SessionTicketExtension.StatelessKey> entry : keyHashMap.entrySet()) {
+            SessionTicketExtension.StatelessKey k = entry.getValue();
+            if (k.isInvalid(this)) {
+                try {
+                    k.key.destroy();
+                } catch (Exception e) {
+                    // Suppress
+                }
+                keyHashMap.remove(entry.getKey());
+            }
+        }
+    }
 
+    protected void addSessionKey(SessionTicketExtension.StatelessKey key) {
+        int newID = key.num;
+        keyHashMap.put(Integer.valueOf(newID), key);
+        currentKeyID = newID;
+        cleanupSessionKeys();
```

Comment:
Does synchronization on sslContext around cleanupSessionKeys is needed? It will block all threads which will try to get the new key, and it seems does not make it safe since the ConcurrentHashMap is used now.